### ISSUE TITLE
Feature Request: define sets of panels depending on file type or project settings

### ISSUE THREAD

**Issue by sboesebeck** [ldelossa/nvim-ide#16]
Hi,

I am working a lot with different projects and different languages (bash, python, java, lua, rust...). So it would be awesome to have the ability to change the default panels depending on the filetype being used. and which panels are opened and which are not.

I would probably try to rely on existing features for this functionality, like adding some code to ftplugin-dir.

e.g. when opening bash-scripts, it would open an explorer on the right, but probably not outline (outline does not really work with lua or bash - maybe another feature request).

when opening a java-file (or related), it would show more panels, maybe by default.

It was mentioned to have some kind of named configurations, so it would be great to use those for example in ftplugins or anything alike...

I hope that does make sense...

**Comment by distek** [comment]
As a workaround, though I think this would be nice to have builtin, you could call `require('ide').setup({...})` in a ftplugin and modify the config as needed for the filetype. However this won't change based on the filetype if you open a new buffer (tested all this just now)

**Comment by ldelossa** [comment]
I like this idea.

**Comment by distek** [comment]
> what really changes in the workspace, since most components are pretty general to "code editing"

I could see this being beneficial if the user's written their own custom component that does something with the particular filetype.

+1 on the "named workspaces"

**Comment by sboesebeck** [comment]
> As a workaround, though I think this would be nice to have builtin, you could call require('ide').setup({...}) in a ftplugin and modify the config as needed for the file type. However this won't change based on the file type if you open a new buffer after opening a different file type (tested all this just now)

yes. This might work for most cases actually. maybe that's the way to go for the time being.

> If its not possible to tie it to file type, would simply having "Named Workspaces" work as well?
> For instance you could start nvim and type "Workspace Bash" to setup a workspace specifically for Bash development?

Yes, that would work - but if the functionality is there, I could also call it from `ftplugin` or in some other lua script - that is exactly my idea. Do not re-invent the wheel - everyone can use the method to their liking and you do not need to implement something.

So, if there is an API for that, e.g. `require("ide").callWorkspace('MyWonderfulConfig')` I could easily put it in all the ftplugin-files I need this functionality for. And if it is possible to switch between them while running, that would be awesome!

**Comment by ldelossa** [comment]
> > As a workaround, though I think this would be nice to have builtin, you could call require('ide').setup({...}) in a ftplugin and modify the config as needed for the file type. However this won't change based on the file type if you open a new buffer after opening a different file type (tested all this just now)
> 
> yes. This might work for most cases actually. maybe that's the way to go for the time being.
> 
> > If its not possible to tie it to file type, would simply having "Named Workspaces" work as well?
> > For instance you could start nvim and type "Workspace Bash" to setup a workspace specifically for Bash development?
> 
> Yes, that would work - but if the functionality is there, I could also call it from `ftplugin` or in some other lua script - that is exactly my idea. Do not re-invent the wheel - everyone can use the method to their liking and you do not need to implement something.
> 
> So, if there is an API for that, e.g. `require("ide").callWorkspace('MyWonderfulConfig')` I could easily put it in all the ftplugin-files I need this functionality for. And if it is possible to switch between them while running, that would be awesome!

Yup this sounds cool. 

Its just a matter of creating the concept and API around "Named Workspaces". 

We can start planning that.

**Comment by sboesebeck** [comment]
cool - I am not a really good lua developer, so I'm probably just of little to no help here. But I am willing to help where I can

**Comment by ldelossa** [comment]
Yup, no worries, I had "Named Workspaces" on the list for features I want to implement.
